### ISSUE TITLE
知识蒸馏模型的teacher模型的预训练参数能不能单独拿出来训练呢？

### ISSUE THREAD

**Issue by OpenOneV** [PaddlePaddle/PaddleOCR#8186]
不知不觉问了四个问题。麻烦万忙中解答了~

1. 从预训练模型pdparams中单独拿出来教师模型可以吗？

2. 如果不方便单独拿出来的话，可不可以通过设置Loss和Architecture中的freeze_params让教师模型单独更新参数而学生模型不参与参数更新呢？
2.1. 最好能关闭学生的前馈，或者让学生不参与最终的Loss计算。不然我只要训练教师模型的话，学生的Loss为什么要算教师头上？这种应该怎么配置Loss和Architecture啊？

3. 请问在yml配置中设置的不同的loss如：
``` yaml
Loss:
  name: CombinedLoss
  loss_config_list:
  - DistillationDilaDBLoss:
      weight: 1.0
      model_name_pairs:
      - ["Student", "Teacher"]
      - ["Student2", "Teacher"]
      key: maps
      balance_loss: true
      main_loss_type: DiceLoss
      alpha: 5
      beta: 10
      ohem_ratio: 3
  - DistillationDMLLoss:
      model_name_pairs:
      - ["Student", "Student2"]
      maps_name: "thrink_maps"
      weight: 1.0
      # model_name_pairs: ["Student", "Student2"]
      key: maps
  - DistillationDBLoss:
      weight: 1.0
      model_name_list: ["Student", "Student2"]
      balance_loss: true
      main_loss_type: DiceLoss
      alpha: 5
      beta: 10
      ohem_ratio: 3
```
这三种loss是最终相加起来的吗？然后交给Optimizer来更新参数吗？

4. 再问一个：PostProcess是在Loss和Opt之间，还是和Opt没有关系啊？

**Comment by LDOUBLEV** [comment]
> 1. 从预训练模型pdparams中单独拿出来教师模型可以吗？

可以，参考文档末尾的代码：https://github.com/PaddlePaddle/PaddleOCR/blob/release/2.6/doc/doc_ch/knowledge_distillation.md


> 如果不方便单独拿出来的话，可不可以通过设置Loss和Architecture中的freeze_params让教师模型单独更新参数而学生模型不参与参数更新呢？
2.1. 最好能关闭学生的前馈，或者让学生不参与最终的Loss计算。不然我只要训练教师模型的话，学生的Loss为什么要算教师头上？这种应该怎么配置Loss和Architecture啊？

如果只想训练teacher参数，参考文档：https://github.com/PaddlePaddle/PaddleOCR/blob/release/2.6/doc/doc_ch/PPOCRv3_det_train.md#22-%E8%AE%AD%E7%BB%83%E6%95%99%E5%B8%88%E6%A8%A1%E5%9E%8B

> 这三种loss是最终相加起来的吗？然后交给Optimizer来更新参数吗？

是的


> 再问一个：PostProcess是在Loss和Opt之间，还是和Opt没有关系啊？

PostProcess 是后处理，不包含参数，不参与Loss计算

**Comment by OpenOneV** [comment]
English search corpus.

Is it OK to separate the teacher model from the pre-trained model pdparams?

Can be a reference document, at the end of the code: https://github.com/PaddlePaddle/PaddleOCR/blob/release/2.6/doc/doc_ch/knowledge_distillation.md



If it is not convenient to take them out separately, can we set Loss and freeze_params in Architecture so that the teacher model can update the parameters independently and the student model does not participate in the parameter update?

2.1. It is better to turn off the feedforward of students, or let students not participate in the final Loss calculation. Otherwise, if I only need to train the teacher model, why should the student's Loss be counted as the teacher's? How should I configure Loss and Architecture?



If you only want to train teacher parameters, refer to the documentation: https://github.com/PaddlePaddle/PaddleOCR/blob/release/2.6/doc/doc_ch/PPOCRv3_det_train.md#22-%E8%AE%AD%E7%BB%83%E6%95%9 9%E5%B8%88%E6%A8%A1%E5%9E%8B



Are these three lossings finally added up? And then turn it over to the Optimizer to update the parameters?



Yes.



One more question: is PostProcess between Loss and Opt, or is it unrelated to Opt?



PostProcess is a post-process that does not include parameters and does not participate in Loss calculation

**Comment by zhangyingbit** [comment]
> > 1. 从预训练模型pdparams中单独拿出来教师模型可以吗？
> 
> 可以，参考文档末尾的代码：https://github.com/PaddlePaddle/PaddleOCR/blob/release/2.6/doc/doc_ch/knowledge_distillation.md
> 
> > 如果不方便单独拿出来的话，可不可以通过设置Loss和Architecture中的freeze_params让教师模型单独更新参数而学生模型不参与参数更新呢？
> > 2.1. 最好能关闭学生的前馈，或者让学生不参与最终的Loss计算。不然我只要训练教师模型的话，学生的Loss为什么要算教师头上？这种应该怎么配置Loss和Architecture啊？
> 
> 如果只想训练teacher参数，参考文档：https://github.com/PaddlePaddle/PaddleOCR/blob/release/2.6/doc/doc_ch/PPOCRv3_det_train.md#22-%E8%AE%AD%E7%BB%83%E6%95%99%E5%B8%88%E6%A8%A1%E5%9E%8B
> 
> > 这三种loss是最终相加起来的吗？然后交给Optimizer来更新参数吗？
> 
> 是的
> 
> > 再问一个：PostProcess是在Loss和Opt之间，还是和Opt没有关系啊？
> 
> PostProcess 是后处理，不包含参数，不参与Loss计算

感谢您的详细解答~
关于识别任务 蒸馏模型的微调：有以下两种方式，(https://github.com/PaddlePaddle/PaddleOCR/blob/18ddb6d5f9bdc2c1b0aa7f6e399ec0f76119dc87/doc/doc_ch/knowledge_distillation.md#215-%E8%92%B8%E9%A6%8F%E6%A8%A1%E5%9E%8B%E5%BE%AE%E8%B0%83)
想问下第一种方式teacher网络会更新吗？如何对teacher网络进行训练呢？谢谢~

**Comment by OpenOneV** [comment]
@zhangyingbit 
直接使用不会，我的做法是：
1. 设置Architecture.Models.Teacher.freeze_params: false
2. 设置Loss.loss_config_list.DistillationDBLoss.model_name_list: ["Student", "Student2", "Teacher"] 从而对Teacher计算和GroundTruth之间的误差。这个新加的误差我估计不会影响两个student模型。而Optimizer则会自动进行梯度下降，也就是不需要操心梯度下降的事情。

![image](https://user-images.githubusercontent.com/115529885/199902143-8ac40699-0af2-4c91-98a5-2648258bdc81.png)
也就是上图给Teacher 也弄一个hard prediction的loss。

测试效果还行。就相当于三个网络一起训练了。不过得调小batch_size我的显存才塞得下。

**Comment by zhangyingbit** [comment]
@OpenOneV 感谢您的详细解答，已经理解了~
